src/form_beam.cpp

```diff
@@ -248,128 +256,101 @@ __global__ void vmBeamform_kernel( gpuDoubleComplex *Jv_Q,
+            int c    = current_beam / n_samples;
+            int nc   = nfine_chan;  /* The (n)umber of (c)hannels (=128) */
+            int s    = current_beam % n_samples; /* The (s)ample number */
+            int ns   = n_samples;  /* The (n)umber of (s)amples (in a chunk)*/
+
+            gpuDoubleComplex ex  = make_gpuDoubleComplex( 0.0, 0.0 );
+            gpuDoubleComplex ey  = make_gpuDoubleComplex( 0.0, 0.0 );
+            gpuDoubleComplex Nxx = make_gpuDoubleComplex( 0.0, 0.0 );
+            gpuDoubleComplex Nxy = make_gpuDoubleComplex( 0.0, 0.0 );
+            gpuDoubleComplex Nyy = make_gpuDoubleComplex( 0.0, 0.0 );
+            // (Nyx is not needed as it's degenerate with Nxy)
+
+            // Just like warps tile beams, threads in a warp tile the antennas.
+            // Each thread computes a partial sum of its corresponding antennas.
+            for(unsigned int ant {lane_id}; ant < nant; ant += warpSize){
+                // Calculate beamform products for each antenna, and then add them together
+                // Calculate the coherent beam (B = J*phi*D)
+                gpuDoubleComplex ex_tmp = gpuCmul( phi[PHI_IDX(p,ant,c,nant,nc)], Jv_Q[Jv_IDX(p,s,c,ant,ns,nc,nant)] );
+                gpuDoubleComplex ey_tmp = gpuCmul( phi[PHI_IDX(p,ant,c,nant,nc)], Jv_P[Jv_IDX(p,s,c,ant,ns,nc,nant)] );
+                ex = gpuCadd(ex, ex_tmp);
+                ey = gpuCadd(ey, ey_tmp);
+                Nxx = gpuCadd(Nxx, gpuCmul( ex_tmp, gpuConj(ex_tmp)));
+                Nxy = gpuCadd(Nxy, gpuCmul( ex_tmp, gpuConj(ey_tmp)));
+                Nyy = gpuCadd(Nyy, gpuCmul( ey_tmp, gpuConj(ey_tmp)));
+            }
+            // intermediate results from thread in a warp are stored in shared memory.
+            workspace[threadIdx.x * 5 + 0] = ex;
+            workspace[threadIdx.x * 5 + 1] = ey;
+            workspace[threadIdx.x * 5 + 2] = Nxx;
+            workspace[threadIdx.x * 5 + 3] = Nxy;
+            workspace[threadIdx.x * 5 + 4] = Nyy;
+
+            // we perform a warp parallel reduction over the partial sums computed by threads in the warp.
+            for(unsigned int i = warpSize/2; i >= 1; i /= 2){
+                if(lane_id < i){
+                    workspace[threadIdx.x * 5] = gpuCadd(workspace[threadIdx.x * 5], workspace[(threadIdx.x + i) * 5]);
+                    workspace[threadIdx.x * 5 + 1] = gpuCadd(workspace[threadIdx.x * 5 + 1], workspace[(threadIdx.x + i) * 5 + 1]);
+                    workspace[threadIdx.x * 5 + 2] = gpuCadd(workspace[threadIdx.x * 5 + 2], workspace[(threadIdx.x + i) * 5 + 2]);
+                    workspace[threadIdx.x * 5 + 3] = gpuCadd(workspace[threadIdx.x * 5 + 3], workspace[(threadIdx.x + i) * 5 + 3]);
+                    workspace[threadIdx.x * 5 + 4] = gpuCadd(workspace[threadIdx.x * 5 + 4], workspace[(threadIdx.x + i) * 5 + 4]);
+                }
+                #ifdef __NVCC__
+                __syncwarp();
```

Comment:
Does there not need to be an equivalent for HIP?

Reply:
That instruction is only needed for NVIDIA GPUs starting from the Volta architecture, that introduces the [independent thread scheduling option](https://stackoverflow.com/questions/70987051/independent-thread-scheduling-since-volta). In such architecture, threads within a warp can execute independently from one another and one of them can "run ahead" of the other ones, possibly creating a race condition.

In AMD GPUs, and NVIDIA GPUs previous to Volta, this is not available. All threads in a warp execute the same instruction in lockstep (or a no-op in thread diverging situation). No thread can run ahead of others in the same warp.

Reply:
Okay. Can you put note there regarding this? Since it may, in future, be a place we need to change some code as the technology develops.

Reply:
Done.